# Week 4 Assignment 5
##### Team 5

### Goal: Select a corpus of choice and analyze high frequency words.

#### 1. Choose a corpus of interest.
The corpus chosen was a collection of quotes from the author Terry Pratchett that can be found here: https://www.lspace.org/books/pqf/

In [74]:
import pandas as pd
import nltk
# nltk.download("all") # You can manually download from nltk by leaving the parentheses empty
from nltk.book import *

In [2]:
# Loading our quotes corpus
f = open("pqf.txt", "rU")
text = f.read()
terrytext = text.split()
quotes = nltk.Text(terrytext)

In [3]:
# Length
len(quotes)

23487

#### 2. How many total unique words are in the corpus? (Please feel free to define unique words in any interesting, defensible way).
Unique was to be defined as words not in the english dictionary. As such, we were to compare the corpus against a dictionary. To achieve this, we first attempted to import the "words" library from nltk.corpus, however, upon discovering that the words that passed as unique mostly included verbs in past or present tense, we decided to rely more on the dictionary of pronounciations from nltk.corpus, since it included verb tenses. Though the results were not perfect, they were better at removing non-unique words. 

In [29]:
# Cleaning up corpus by removing punctuation and common words
from nltk.corpus import stopwords
quotes = [w for w in quotes if w.isalpha()]
quotes = [w for w in quotes if w not in stopwords.words('english')]

In [50]:
# Unique words from pronounciation syllables
entries = nltk.corpus.cmudict.entries()
entries = [item[0] for item in entries] # Remove pronounciation
len(entries)

133737

In [52]:
# Remove words not found in our pronounciation dictionary
# Because the words weren't made lowercase earlier in the "quotes" list, we do it again to remove common words
sub  = [w.lower() for w in quotes if w not in entries]
sub = [w for w in sub if w not in stopwords.words('english')]

In [61]:
sub[0:9]

['surely', 'people', 'star', 'trek', 'ye', 'cannae', 'oout', 'feegle', 'wee']

In [62]:
len(sub)

1843

#### 3. Taking the most common words, how many unique words represent half of the total words in the corpus?
Since our corpus contained 23,487 words, the number of most common unique words that take 50% of the corpus should be calculate in such a way that summing the count of each of the unique words adds up to approximately 11,744 words. Becaue the pool of unique words is so small, the sum of the words totals 1,164, which is nowhere near 50% of corpus' volume. At best, it is 5%.

In [73]:
topunique = nltk.FreqDist(sub)
topcount = pd.DataFrame(topunique.most_common(200), columns=['word','n'])
topcount.head(10)

word   n
0      small  54
1       good  53
2  pratchett  51
3       neil  51
4     moving  51
5      lords  34
6      death  26
7    gaspode  24
8        men  22
9     johnny  21

In [72]:
Total = topcount['n'].sum()
print (Total)

1164


#### 4. Identify the 200 highest frequency words in this corpus.

In [77]:
# build the frequency distribution using FreqDist()
quotes_freq = nltk.FreqDist(quotes)
quotes_freq = pd.DataFrame(quotes_freq.most_common(200), columns=['word','n'])
quotes_freq['relative_frequency'] = quotes_freq['n']/float(len(quotes))
quotes_freq.head(20)

word    n  relative_frequency
0         The  197            0.021594
1           I  139            0.015236
2        said   93            0.010194
3         one   82            0.008988
4      people   73            0.008002
5          It   68            0.007454
6        like   68            0.007454
7       Small   54            0.005919
8        Good   52            0.005700
9      Moving   51            0.005590
10       Neil   51            0.005590
11  Pratchett   51            0.005590
12      could   50            0.005481
13          A   47            0.005152
14        get   43            0.004713
15      would   38            0.004165
16     always   37            0.004056
17        And   37            0.004056
18    thought   37            0.004056
19      Lords   34            0.003727

#### 5. Create a graph that shows the relative frequency of these 200 words.

#### 6. Does the observed relative frequency of these words follow Zipf’s law? Explain.

In [6]:
# Zipf's law: given a large sample of words used, the frequency of any word is inversely
# proportional to its rank in the frequency table. Aka, low rank on the frequency table
# means very frequently used word. The most frequent word will occur approximately twice
# as often as the second most frequent word, which will occur approximately twice as often
# as the fourth most frequent word 

#### 7. In what ways do you think the frequency of the words in this corpus differ from “all words in all corpora.”